Issue/8340 add history detail (Part 1)

theck13 opened the pull request:
### Fix
Add a detailed view of a version selected from the ***History*** list in the [full-screen dialog](https://material.io/design/components/dialogs.html#full-screen-dialog) format with [FullScreenDialogFragment](https://github.com/wordpress-mobile/WordPress-Android/blob/f9b84df81c47bfa6278df1342b035f811d3412be/WordPress/src/main/java/org/wordpress/android/ui/FullScreenDialogFragment.java) including a ***Load*** action in the top toolbar, additions count in the bottom toolbar, and deletions count in the bottom toolbar as described in https://github.com/wordpress-mobile/WordPress-Android/issues/8340.

![revisions_history_detail](https://user-images.githubusercontent.com/3827611/47217509-8ffb1a80-d376-11e8-9ec4-a50b1caf3f24.png)

### Note
There are a couple differences between the ***Design*** and ***Develop*** screenshots.  The toolbar color was changed from `grey_dark`(`#2e4453`) to `blue_wordpress`(`#0087be`).  Since the status bar color cannot be changed, using `blue_wordpress` works better with the app's status bar color (`#006b98`) and is consistent with the other full-screen dialog (i.e. ***Username Changer***). The toolbar subtitle does not have the revision status following the time ago.  The API does not provide the page/post status for each revision.  The bottom bar chevrons and swiping left/right to move between versions are not included.  Those features will be in a subsequent pull request in order to make two smaller changesets rather than one larger changeset for easier reviewing.  The image is not rendered and the HTML is shown.  That allows for a more detailed view of differences to HTML elements (e.g. `src`, `height`, `width`), which would otherwise be hidden.

### Test
#### Site Pages
1. Go to ***My Site*** tab.
2. Tap ***Site Pages*** item under ***Publish*** section.
3. Tap page in ***Site Pages*** list.
4. Tap overflow menu button in top toolbar.
5. Tap ***History*** item in menu.
6. Tap item in ***History*** list.
7. Notice ***Revision*** full-screen dialog shown above.
8. Tap ***X*** action.
9. Notice ***Revision*** full-screen dialog is dismissed.
10. Tap item in ***History*** list.
11. Notice ***Revision*** full-screen dialog shown above.
12. Tap ***Load*** action.
13. Notice loading progress dialog shown above.
14. Notice revision title and content loaded.
15. Notice snackbar with ***Undo*** action shown above.
16. Tap ***Undo*** action in snackbar.
17. Notice revision title and content are undone.

#### Blog Posts
1. Go to ***My Site*** tab.
2. Tap ***Blog Posts*** item under ***Publish*** section.
3. Tap post in ***Blog Posts*** list.
4. Tap overflow menu button in top toolbar.
5. Tap ***History*** item in menu.
6. Tap item in ***History*** list.
7. Notice ***Revision*** full-screen dialog shown above.
8. Tap ***X*** action.
9. Notice ***Revision*** full-screen dialog is dismissed.
10. Tap item in ***History*** list.
11. Notice ***Revision*** full-screen dialog shown above.
12. Tap ***Load*** action.
13. Notice loading progress dialog shown above.
14. Notice revision title and content loaded.
15. Notice snackbar with ***Undo*** action shown above.
16. Tap ***Undo*** action in snackbar.
17. Notice revision title and content are undone.

### Review
Only one developer and one designer are required to review these changes, but anyone can perform the review.

theck13 commented:
> I noticed that title gets ellipsized if it's too long, so I wonder, should we allow multiline (with some limint, say 5 lines?) and just trim the newline at the end?

It looks like the title field isn't limited at all in the editor so we should probably mimic that behavior with the history detail view.  That means removing the [`android:singleLine="true"`](https://github.com/wordpress-mobile/WordPress-Android/pull/8457/files#diff-c8ec817b105c99917677e7e4b6c212eeR81) attribute from the title view in `history_detail_dialog_fragment` and trimming the newline at the end.  Do you think we should do that in `DiffView.showDiffs`?

> Also, I experienced a crash when you minimize the app while at the Details screen:

I was able to reproduce that crash.  Thanks for catching it!  I'll fix that.

khaykov commented:
> It looks like the title field isn't limited at all in the editor so we should probably mimic that behavior with the history detail view. That means removing the android:singleLine="true" attribute from the title view in history_detail_dialog_fragment and trimming the newline at the end. Do you think we should do that in DiffView.showDiffs?

Doing it in `DiffView.showDiffs` sounds good to me, if we implement it in some generic way, like `DiffView.showDiffs(diffs: List<Diff>, trimDiff: Boolean = false)`. Newline might still be a legit change in post content.

theck13 commented:
I removed the `singleLine` attribute (a40b529), updated the `DiffView.showDiffs` method (e96678a), and updated the FluxC dependency (4813295).  It's ready for another code review!

khaykov commented on WordPress/src/main/java/org/wordpress/android/widgets/DiffView.kt:
```diff
@@ -18,10 +18,14 @@ class DiffView : AppCompatTextView {
         text = null
 
         diffs.forEachIndexed { index, diff ->
+            if (trimNewline && index == diffs.size - 1 && diff.value.equals("\n")) {
```
After some testing, looks like there are some instances where the newline is not the only element in the trailing diff, so we might want to use something like `trimEnd('\n')` on the last diff value.

theck13 replied:
I wasn't able to reproduce a `Diff` with anything other than a single `\n` character, but I updated the `DiffView.showDiffs` method in ab20d08.

khaykov commented:
@theck13 Thanks for the fixes! Looks good to me :)

iamthomasbishop commented:
Looking at the screenshots above to review, they look great overall! A couple minor things:

* Is there any way we can tighten up the spacing between the title and subtitle on the app bar? Maybe this would mean negative spacing there, but it feels kind of wide there

* I understand we're not changing the app bar color to `grey-dark` because it clashes with the status bar color. Can we try changing to the same color as the status bar? That might be enough to achieve the desired effect

> The image is not rendered and the HTML is shown. That allows for a more detailed view of differences to HTML elements (e.g. src, height, width), which would otherwise be hidden.

Is there any way to work around this? It's not ideal to show HTML because the idea is to show an accurate (enough) representation visually so they can make a connection with what changed. I think an earlier iteration had allowed the user to switch to HTML Mode in this view, but we decided against it in favor of focusing on a visual representation.

khaykov commented:
> Is there any way to work around this? It's not ideal to show HTML because the idea is to show an accurate (enough) representation visually so they can make a connection with what changed. I think an earlier iteration had allowed the user to switch to HTML Mode in this view, but we decided against it in favor of focusing on a visual representation.

@iamthomasbishop  Maybe I'm missing some simple solution, but in my understanding, visual diffs are pretty challenging. We need to be able to show changes that have a very subtle visual effect (like changing some attribute of an image or style) or no visual effect at all. One way this could potentially work is with side by side comparison, but this might be tricky on mobile.

Do you have any examples of apps/editors/software that has this functionality with rich content?

theck13 commented:
> What font size is that subtitle at? It might be worth shrinking it down if it brings the bottom edge up toward the title and is still readable (12-13pt+). Also, do we have the ability to make the subtitle white color as well?

The default subtitle size is 16sp, which is the standard for type [Subtitle 1 and Body 1](https://material.io/design/typography/the-type-system.html#type-scale).  I updated the size to 14sp and color to white in 80624ea, which is the standard for type [Subtitle 2 and Body 2](https://material.io/design/typography/the-type-system.html#type-scale).  See the screenshots below for illustration.

![revisions_history_detail_subtitle](https://user-images.githubusercontent.com/3827611/47463956-be815700-d7b5-11e8-818b-6a3ef4fa2e1f.png)

iamthomasbishop commented:
Awesome, lgtm! :shipit:
